packages/langium/src/lsp/semantic-token-provider.ts

```diff
@@ -0,0 +1,188 @@
+/******************************************************************************
+ * Copyright 2022 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+/* eslint-disable no-bitwise */
+
+import { CancellationToken, Range, SemanticTokenModifiers, SemanticTokens, SemanticTokensBuilder, SemanticTokensDelta, SemanticTokensDeltaParams, SemanticTokensOptions, SemanticTokensParams, SemanticTokensRangeParams, SemanticTokenTypes } from 'vscode-languageserver';
+import { findKeywordNode, findNodeForFeature } from '../grammar/grammar-util';
+import { AstNode, CstNode, Properties } from '../syntax-tree';
+import { AstNodeContent, streamAllContents } from '../utils/ast-util';
+import { LangiumDocument } from '../workspace/documents';
+
+export const AllSemanticTokenTypes: Record<string, number> = {
+    [SemanticTokenTypes.class]: 0,
+    [SemanticTokenTypes.comment]: 1,
+    [SemanticTokenTypes.enum]: 2,
+    [SemanticTokenTypes.enumMember]: 3,
+    [SemanticTokenTypes.event]: 4,
+    [SemanticTokenTypes.function]: 5,
+    [SemanticTokenTypes.interface]: 6,
+    [SemanticTokenTypes.keyword]: 7,
+    [SemanticTokenTypes.macro]: 8,
+    [SemanticTokenTypes.method]: 9,
+    [SemanticTokenTypes.modifier]: 10,
+    [SemanticTokenTypes.namespace]: 11,
+    [SemanticTokenTypes.number]: 12,
+    [SemanticTokenTypes.operator]: 13,
+    [SemanticTokenTypes.parameter]: 14,
+    [SemanticTokenTypes.property]: 15,
+    [SemanticTokenTypes.regexp]: 16,
+    [SemanticTokenTypes.string]: 17,
+    [SemanticTokenTypes.struct]: 18,
+    [SemanticTokenTypes.type]: 19,
+    [SemanticTokenTypes.typeParameter]: 20,
+    [SemanticTokenTypes.variable]: 21
+};
+
+export const AllSemanticTokenModifiers: Record<string, number> = {
+    [SemanticTokenModifiers.abstract]: 1 << 0,
+    [SemanticTokenModifiers.async]: 1 << 1,
+    [SemanticTokenModifiers.declaration]: 1 << 2,
+    [SemanticTokenModifiers.defaultLibrary]: 1 << 3,
+    [SemanticTokenModifiers.definition]: 1 << 4,
+    [SemanticTokenModifiers.deprecated]: 1 << 5,
+    [SemanticTokenModifiers.documentation]: 1 << 6,
+    [SemanticTokenModifiers.modification]: 1 << 7,
+    [SemanticTokenModifiers.readonly]: 1 << 8,
+    [SemanticTokenModifiers.static]: 1 << 9
+};
```

Comment:
Strange that this is not part of `vscode-languageserver`.

Reply:
Usually a language server should only add those modifiers and types to the token legend, which are actually in use. Additionally, you can actually add custom modifiers and types, which are special to your IDE. I imagine there's some optimization going on, which is why they probably don't want to export some kind of "catch-all" token legend. In my first iteration the semantic token provider actually had a method to return all used types and modifiers, but that proved to be a lot of boilerplate code and really complex for multiple token providers.